guardian/support-dotcom-components
# Parse Epic tests coming from dynamo

## What does this change? [Trello](https://trello.com/c/cCe7LieX/148-sdc-handle-invalid-test-config-gracefully)

Parse epic tests coming from dynamo DB with zod schemas - if a test config is not successfully parsed we raise an alarm and filter it out. This comes after multiple occasions where instances have failed the health check because of badly configured dynamo data. 

We've kept separately defined TS types and zod schemas (inferring the types from the schemas will be [future work](https://trello.com/c/rG9ooh7N/134-consider-infering-sdc-types-from-zod-schemas)). However the types are aligned by using a generic `T` that is validated by `schema?: ZodSchema<T>` (eg `T` is `EpicTest` and schema is `EpicTestSchema`). See potential risks for a limitation of this.



## How to test

Plan:
Deploy to CODE - instances should be healthy. We can manually break the config and check that it raises an alarm and starts up healthily. 

Result:
We have done this - alarm was raised and error logs showed up in Kibana.

Plan: 
Epics should render as normal in CODE DCR and RRCP

Result:
Yes



## How can we measure success?



## Have we considered potential risks?

One limitation of keeping separate Typescript types and zod schemas is that optional fields are not passed through validation if they are not specified in the schema.

Eg:

Suppose we have the type:
`
type Example = { 
    foo: string;
    bar?: string;
 }
`

and the Zod schema:

`
const exampleSchema = z.object({
    foo: z.string();
});
`

if we parse the JSON object: 
`{
   foo: 'foo';
   bar: 'bar';
}`

using `exampleSchema`, `bar` will not be present in the resulting data, but it would satisfy the `ZodSchema<T>` relationship where `T` is `Example`. The mitigation to this is to manually check that our types and schemas match for optional properties.



## Images



## Accessibility
